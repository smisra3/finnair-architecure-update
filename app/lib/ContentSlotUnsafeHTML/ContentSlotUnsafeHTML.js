// @flow
import type { Node } from 'react';
import { PureComponent } from 'react';
import unescape from 'lodash/unescape';
import memoizeLast from '../../utils/memoizeLast';

type Props = {
  content: string,
  hasScript?: boolean,
  wrapper?: Node,
};

/**
 * Component to insert the HTML content from the ATG BCC layer
 * which is authorable by the content authors (product team)
 */
class ContentSlotUnsafeHTML extends PureComponent<Props> {
  static defaultProps = {
    hasScript: true,
    wrapper: null,
  };

  static scriptRegEx = /(<script\b[^>]*>[\s\S]*?<\/script>)/gm;

  // Regex to replace HTML ASCII code &#034 (coming as escaped HTML from server code) to `"`
  static htmlDecode = (input: string = '') =>
    (input && input.replace ? unescape(input.replace(/&#034;/gi, '"')) : '');

  static createMarkupForBody = memoizeLast(contentSlotUnsafeHTML => ({
    __html: ContentSlotUnsafeHTML.htmlDecode(contentSlotUnsafeHTML).replace(
      ContentSlotUnsafeHTML.scriptRegEx,
      '',
    ),
  }));

  static extractScripts = (input: string) => {
    let match = ContentSlotUnsafeHTML.scriptRegEx.exec(input);
    const scriptList: Array<string> = [];
    while (match) {
      scriptList.push(match[0]);
      match = ContentSlotUnsafeHTML.scriptRegEx.exec(input);
    }
    return scriptList;
  };

  componentDidMount = () => {
    if (this.props.hasScript && document && document.createRange) {
      const { content } = this.props;
      const scriptList = ContentSlotUnsafeHTML.extractScripts(content);
      const range = document.createRange();
      range.setStart(document.body, 0);
      document.body.appendChild(range.createContextualFragment(scriptList.join('')));
    }
  };
  render(): Node {
    const {
      content, wrapper, hasScript, ...other
    } = this.props;
    /* eslint-disable react/no-danger */
    const Wrapper = wrapper;
    return wrapper ? (
      <Wrapper>
        <div
          dangerouslySetInnerHTML={ContentSlotUnsafeHTML.createMarkupForBody(content)}
          {...other}
        />
      </Wrapper>
    ) : (
      <div
        dangerouslySetInnerHTML={ContentSlotUnsafeHTML.createMarkupForBody(content)}
        {...other}
      />
    );
    /* eslint-enable react/no-danger */
  }
}

export default ContentSlotUnsafeHTML;
