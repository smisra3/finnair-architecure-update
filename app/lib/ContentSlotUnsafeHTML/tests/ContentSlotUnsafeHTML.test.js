import { shallow } from 'enzyme';
import ContentSlotUnsafeHTML from '../';

/* eslint-disable import/prefer-default-export */
export const props = {
  content: '<div>HTML</div>',
  hasScript: true,
};

describe('ContentSlotUnsafeHTML Component', () => {
  let contentSlot = '';

  Object.defineProperty(document, 'createRange', {
    value() {
      return {
        setStart() {},
        createContextualFragment() {},
      };
    },
  });
  Object.defineProperty(document.body, 'appendChild', {
    value() {},
  });

  beforeEach(() => {
    contentSlot = shallow(<ContentSlotUnsafeHTML {...props} />);
  });

  test('should render correctly', () => {
    expect(contentSlot).toMatchSnapshot();
  });

  test('should render correclty without Wrapper', () => {
    props.wrapper = '';
    const contentSlotMount = shallow(<ContentSlotUnsafeHTML {...props} />);
    expect(contentSlotMount).toMatchSnapshot();
  });

  test('should render correclty with Wrapper', () => {
    props.wrapper = 'div';
    const contentSlotMount = shallow(<ContentSlotUnsafeHTML {...props} />);
    expect(contentSlotMount).toMatchSnapshot();
  });
});
