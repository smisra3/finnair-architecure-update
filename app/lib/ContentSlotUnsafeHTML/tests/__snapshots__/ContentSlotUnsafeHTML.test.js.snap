// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentSlotUnsafeHTML Component should render correclty with Wrapper 1`] = `
<div>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<div>HTML</div>",
      }
    }
  />
</div>
`;

exports[`ContentSlotUnsafeHTML Component should render correclty without Wrapper 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<div>HTML</div>",
    }
  }
/>
`;

exports[`ContentSlotUnsafeHTML Component should render correctly 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<div>HTML</div>",
    }
  }
/>
`;
